### PR TITLE
feat(ingest): snowflake using oauth

### PR DESCRIPTION
**Pull request by saxo-lalrishav**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by maggiehays** [comment]
Hi @anshbansal please pick up review on this one -- let's see if we can replicate on our Snowflake instance

**Comment by maggiehays** [comment]
Hi @anshbansal please pick up review on this one -- let's see if we can replicate on our Snowflake instance

**Comment by anshbansal** [comment]
@saxo-lalrishav Few comments
- Please follow commit message format https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format 
- Please resolve conflicts with master
- Please fix the lint. You might need to re-create your local venv to get the latest updates of lint libraries

I'll try and see see if I can validate this PR on my local.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        self.client_id = client_id
+        self.authority_url = authority_url
+
+    def _get_token(
+        self,
+        credentials: Union[str, Dict[str, Any]],
+        scope: Union[str, List[str]],
+        check_cache: bool,
+    ) -> str:
+        app = msal.ConfidentialClientApplication(
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:22]: This makes it Microsoft specific. Can we please change the `OauthTokenGenerator` to have some options for future extensibility when the 2 other cloud providers support is added? Please keep this Microsoft specific caveat in mind while writing docs.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        )
+
+        CLIENT_CREDENTIAL = {
+            "thumbprint": public_cert_thumbprint,
+            "private_key": decoded_private_key_content,
+        }
+        return self._get_token(CLIENT_CREDENTIAL, scope, check_cache)
+
+    def get_token_with_secret(
+        self, secret: str, scope: Union[str, List[str]], check_cache: bool = False
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:61]: Can we please change `scope` to be `List[str]` only? And make the corresponding changes? I don't see why we need to have it a Union of both. Wherever we are passing we can make it `List[str]`.

`scopes` might represent that better.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    def get_public_certificate_thumbprint(self, public_cert_str: str) -> str:
+        cert_str = public_cert_str
+        certificate = load_certificate(FILETYPE_PEM, cert_str.encode("utf-8"))
+        return (certificate.digest("sha1")).decode("utf-8").replace(":", "")
+
+    def get_token_with_certificate(
+        self,
+        private_key_content: str,
+        public_key_content: str,
+        scope: Union[str, List[str]] = "https://management.core.windows.net/.default",
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:45]: This default seems very arbitrary. Please pass it to `OauthTokenGenerator` via the config option for Microsoft specific things.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        if check_cache:
+            token = app.acquire_token_silent(scopes=scope, account=None)
+
+        if not token:
+            log.debug("No token available in cache. Getting a new one from AAD.")
+            token = app.acquire_token_for_client(scopes=scope)
+        return token
+
+    def get_public_certificate_thumbprint(self, public_cert_str: str) -> str:
+        cert_str = public_cert_str
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:37]: Why is this here?

File: metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py [hunk 5]
```diff
@@ -69,6 +73,13 @@ class BaseSnowflakeConfig(BaseTimeWindowConfig):
 
     scheme: str = "snowflake"
     username: Optional[str] = None
+    use_certificate: Optional[bool] = False
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py:76]: Should we extract all these `oauth` settings in their own object? Currently it is hard to know what is required. 3 of them are not optional but this makes it hard to see whether they are optional or not.

File: metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py [hunk 6]
```diff
@@ -104,6 +115,39 @@ def authenticator_type_is_valid(cls, v, values, **kwargs):
...
+                if values.get("oauth_authority_url") is None:
+                    raise ValueError(
+                        f"'oauth_authority_url' was none "
+                        f"but should be set when using {v} authentication"
+                    )
+                if values.get("use_certificate") is True:
+                    if values.get("base64_encoded_oauth_private_key") is None:
+                        raise ValueError(
+                            f"'base64_encoded_oauth_private_key' was none "
+                            f"but should be set when using {v} authentication"
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py:138]: This message needs to be changed and add it is needed when `use_certificate` is true

File: metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py [hunk 7]
```diff
@@ -104,6 +115,39 @@ def authenticator_type_is_valid(cls, v, values, **kwargs):
...
+                        )
+                    if values.get("base64_encoded_oauth_public_key") is None:
+                        raise ValueError(
+                            f"'base64_encoded_oauth_public_key' was none"
+                            f"but should be set when using {v} authentication"
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py:143]: This message needs to be changed and add it is needed when `use_certificate` is true

File: metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py [hunk 8]
```diff
@@ -104,6 +115,39 @@ def authenticator_type_is_valid(cls, v, values, **kwargs):
...
+                        )
+                else:
+                    if values.get("oauth_client_secret") is None:
+                        raise ValueError(
+                            f"'oauth_client_secret' was none "
+                            f"but should be set when using {v} authentication"
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py:149]: This message needs to be changed and add it is needed when `use_certificate` is true

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py (hunk 3 above)
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:45]: I guess we should be able to infer from the `host_name` whether this is Microsoft or not. We can add a warning to the source when `Oauth` is being used but snowflake is not on Microsoft.

**Review by anshbansal — CHANGES_REQUESTED** [review]
Needs changes. Please also add docs.

**Comment by anshbansal** [comment]
@saxo-lalrishav Is this something that you are still working on?

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
+import base64
+import logging
+from typing import Any, Dict, List, Union
+
+import msal
```
[the remark this answers is not in the document]
Comment (saxo-lalrishav) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:5]: this is already mentioned here - https://github.com/datahub-project/datahub/blob/23f657ed76a30e2521bdd28494201e89e289cb49/metadata-ingestion/setup.py#L116

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py (hunk 1 above)
Comment (saxo-lalrishav) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:31]: added different functions for each provider, based on provider passed in the recipe
right now get_token function is only implemented for microsoft

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py (hunk 2 above)
Comment (saxo-lalrishav) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py]: done

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py (hunk 3 above)
Comment (saxo-lalrishav) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py]: remove the deafult value

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py (hunk 4 above)
Comment (saxo-lalrishav) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:51]: this `get_public_certificate_thumbprint` function is used to fetch thumbprinbt of the certificate, which is later used by msal library to generate the toke is use_certificate is true

File: metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py (hunk 5 above)
Comment (saxo-lalrishav) [reply metadata-ingestion/src/datahub/ingestion/source_config/sql/snowflake.py]: done

File: settings.gradle [hunk 10]
```diff
@@ -44,4 +44,4 @@ include 'entity-registry:custom-test-model'
 include 'metadata-integration:java:spark-lineage'
 include 'metadata-integration:java:datahub-client'
 include 'metadata-integration:java:datahub-protobuf'
-include 'ingestion-scheduler'
```
Comment (anshbansal) [inline settings.gradle:47]: Please revert this change.

File: metadata-ingestion/archived/source_docs/snowflake.md [hunk 11]
```diff
@@ -170,7 +170,7 @@ Note that a `.` is used to denote nested fields in the YAML recipe.
```
Comment (anshbansal) [inline metadata-ingestion/archived/source_docs/snowflake.md:170]: Docs have moved into source code.  Please see latest snowflake config on master and merge the changes in.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oauth_generator.py:5]: This is not added in `snowflake` only powerbi. This might cause snowflake connector to fail for people not using powerbi

File: metadata-ingestion/archived/source_docs/snowflake.md (hunk 11 above)
Comment (saxo-lalrishav) [reply metadata-ingestion/archived/source_docs/snowflake.md]: done

**Comment by saxo-lalrishav** [comment]
@anshbansal pls have a look into this and provide your feedback

**Comment by anshbansal** [comment]
https://github.com/datahub-project/datahub/pull/4647#discussion_r866049376 This comment is remaining
Please update description from "oauth configuration" to add the snowflake doc where the details are present.
Please merge latest master so that linting passes.
And please add an example configuration at least in the PR so we know what the whole config looks like for snowflake for microsoft. A better thing to do would be to add unit tests for the big validator that you have added.

**Comment by anshbansal** [comment]
Hi @saxo-lalrishav Are you still working on this PR? We can merge it after the changes requested in last comment.

**Comment by saxo-lalrishav** [comment]
> Hi @saxo-lalrishav Are you still working on this PR? We can merge it after the changes requested in last comment.

Yeah, looking into this

**Comment by anshbansal** [comment]
For the lint failures suggest to re-create your local venv from scratch so that you have the latest libraries.

File: metadata-ingestion/tests/unit/test_snowflake_source.py [hunk 12]
```diff
@@ -14,6 +14,51 @@ def test_snowflake_source_throws_error_on_account_id_missing():
         )
 
 
+def test_snowflake_throws_error_on_client_id_missing_if_using_oauth():
+    with pytest.raises(ConfigurationError):
+        SnowflakeConfig.parse_obj(
+            {
+                "authentication_type": "OAUTH_AUTHENTICATOR",
+                "oauth_config":
```
[the remark this answers is not in the document]
Comment (saxo-lalrishav) [reply metadata-ingestion/tests/unit/test_snowflake_source.py]: Ohh sorry , i got it, fixing it

**Review by anshbansal — CHANGES_REQUESTED** [review]
This looks like a bug in the unit tests. Are you running these unit tests on local while testing? Please see https://datahubproject.io/docs/metadata-ingestion/developing/#sanity-check-code-before-committing

```
>               if values.get("oauth_config").provider is None:
E               AttributeError: 'NoneType' object has no attribute 'provider'

src/datahub/ingestion/source_config/sql/snowflake.py:192: AttributeError
_ test_snwoflake_throws_error_on_encoded_oauth_private_key_missing_if_use_certificate_is_true _

    def test_snwoflake_throws_error_on_encoded_oauth_private_key_missing_if_use_certificate_is_true():
        with pytest.raises(ConfigurationError):
            SnowflakeConfig.parse_obj(
                {
                    "authentication_type": "OAUTH_AUTHENTICATOR",
                    "oauth_config": {
                        "client_id": "882e9831-7ea51cb2b954",
                        "provider": "microsoft",
                        "scopes": "[https://microsoft.com/f4b353d5-ef8d/.default]",
                        "use_certificate": True,
                        "authority_url": "https://login.microsoftonline.com/yourorganisation.com",
>                       "encoded_oauth_public_key": "fkdsfhkshfkjsdfiuwrwfkjhsfskfhksjf==",
                    },
                }
            )

```

**Review by anshbansal — APPROVED** [review]
LGTM
